### PR TITLE
Added support for analytics_label

### PR DESCRIPTION
**Pull request by RannyRanny**
Solves https://github.com/firebase/firebase-admin-dotnet/issues/88

### CODE REVIEW

**Comment by RannyRanny** [comment]
I signed it!

File: FirebaseAdmin/FirebaseAdmin/Messaging/AndroidConfig.cs [hunk 1]
```diff
@@ -66,6 +66,12 @@ public sealed class AndroidConfig
         [JsonProperty("notification")]
         public AndroidNotification Notification { get; set; }
 
+        /// <summary>
+        /// Gets or sets the FCM options to be included in the message.
+        /// </summary>
+        [JsonProperty("fcm_options")]
+        public FcmOptions FcmOptions { get; set; }
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/AndroidConfig.cs]: We need a new `AndroidFcmOptions` type here. See how this was implemented in Java: https://github.com/firebase/firebase-admin-java/pull/285

File: FirebaseAdmin/FirebaseAdmin/Messaging/ApnsConfig.cs [hunk 2]
```diff
@@ -35,6 +35,12 @@ public sealed class ApnsConfig
         [JsonProperty("headers")]
         public IReadOnlyDictionary<string, string> Headers { get; set; }
 
+        /// <summary>
+        /// Gets or sets the FCM options to be included in the message.
+        /// </summary>
+        [JsonProperty("fcm_options")]
+        public FcmOptions FcmOptions { get; set; }
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/ApnsConfig.cs]: Introduce a new `ApnsFcmOptions` type.

File: FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+using Newtonsoft.Json;
+
+namespace FirebaseAdmin.Messaging
+{
+    /// <summary>
+    /// Represents FCM options.
+    /// </summary>
+    public class FcmOptions
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs]: sealed

File: FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        /// <summary>
+        /// Gets or sets analytics label.
+        /// </summary>
+        [JsonProperty("analytics_label")]
+        public string AnalyticsLabel { get; set; }
+    }
+}
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs]: Add newline

**Comment by RannyRanny** [comment]
@hiranya911 Added new classes and coveder tests

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs [hunk 5]
```diff
@@ -1650,6 +1685,42 @@ public void WebpushNotificationWithInvalidHttpsLinkUrl()
...
+                    }
+                },
+                {
+                    "fcm_options", new JObject()
+                    {
+                        { "analytics_label", "label" },
+                    }
+                },
+            };
+            Assert.Throws<ArgumentException>(() => this.AssertJsonEquals(expected, message));
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs]: Can we just call `message.CopyAndValidate()` and get rid of `expected`?

File: FirebaseAdmin/FirebaseAdmin/Messaging/AndroidFcmOptions.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        /// Copies this FCM options, and validates the content of it to ensure that it can
+        /// be serialized into the JSON format expected by the FCM service.
+        /// </summary>
+        internal AndroidFcmOptions CopyAndValidate()
+        {
+            var copy = new AndroidFcmOptions()
+            {
+                AnalyticsLabel = this.AnalyticsLabel,
+            };
+            AnalyticsLabelChecker.CheckAnalyticsLabelOrThrow(this.AnalyticsLabel);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/AndroidFcmOptions.cs]: Pass `copy.AnalyticsLabel` to the validation function.

File: FirebaseAdmin/FirebaseAdmin/Messaging/ApnsFcmOptions.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        /// Copies this FCM options, and validates the content of it to ensure that it can
+        /// be serialized into the JSON format expected by the FCM service.
+        /// </summary>
+        internal ApnsFcmOptions CopyAndValidate()
+        {
+            var copy = new ApnsFcmOptions()
+            {
+                AnalyticsLabel = this.AnalyticsLabel,
+            };
+            AnalyticsLabelChecker.CheckAnalyticsLabelOrThrow(this.AnalyticsLabel);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/ApnsFcmOptions.cs]: `copy.AnalyticsLabel`

File: FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        /// Copies this FCM options, and validates the content of it to ensure that it can
+        /// be serialized into the JSON format expected by the FCM service.
+        /// </summary>
+        internal FcmOptions CopyAndValidate()
+        {
+            var copy = new FcmOptions()
+            {
+                AnalyticsLabel = this.AnalyticsLabel,
+            };
+            AnalyticsLabelChecker.CheckAnalyticsLabelOrThrow(this.AnalyticsLabel);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs]: `copy.AnalyticsLabel`

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace FirebaseAdmin.Messaging.Util
+{
+    /// <summary>
+    /// Checker for analytics label.
+    /// </summary>
+    public static class AnalyticsLabelChecker
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: internal

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        private static ImmutableHashSet<char> alowedChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-~%".ToCharArray().ToImmutableHashSet();
+
+        /// <summary>
+        /// Checks anytytics labels an throw if not valid.
+        /// </summary>
+        /// <exception cref="ArgumentException">If analytics label does not match pattern.</exception>
+        /// <param name="analyticsLabel">Analytics label.</param>
+        public static void CheckAnalyticsLabelOrThrow(string analyticsLabel)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: Nit: Perhaps just call it `ValidateAnalyticsLabel()`. I don't feel too strongly about it, so your call.

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            foreach (var character in analyticsLabel)
+            {
+                if (!alowedChars.Contains(character))
+                {
+                    throw new ArgumentException("Analytics label must have format matching'^[a-zA-Z0-9-_.~%]{1,50}$");
+                }
+            }
+        }
+    }
+}
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: Add newline.

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        private static ImmutableHashSet<char> alowedChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-~%".ToCharArray().ToImmutableHashSet();
+
+        /// <summary>
+        /// Checks anytytics labels an throw if not valid.
+        /// </summary>
+        /// <exception cref="ArgumentException">If analytics label does not match pattern.</exception>
+        /// <param name="analyticsLabel">Analytics label.</param>
+        public static void CheckAnalyticsLabelOrThrow(string analyticsLabel)
+        {
+            if (analyticsLabel.Length > 50)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: Can we just do a regex match like in Java?

```
Regex.IsMatch(analyticsLabel, pattern)
```

You should also handle the null case.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs [hunk 13]
```diff
@@ -75,6 +75,10 @@ public void Notification()
                     Title = "title",
                     Body = "body",
                 },
+                FcmOptions = new FcmOptions()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs]: Can you also add a test case to see what happens when the AnalyticsLabel is unspecified (i.e. null)? I think the current implementation will fail in that case.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs [hunk 14]
```diff
@@ -1650,6 +1685,42 @@ public void WebpushNotificationWithInvalidHttpsLinkUrl()
...
+            {
+                Topic = "test-topic",
+                Notification = new Notification()
+                {
+                    Title = "title",
+                    Body = "body",
+                },
+                FcmOptions = new FcmOptions()
+                {
+                    AnalyticsLabel = "label!",
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/MessageTest.cs]: Add additional test cases for the cases when the label is empty and longer than 50.

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs (hunk 12 above)
Comment (RannyRanny) [reply FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: `Regex.IsMatch(analyticsLabel, pattern)` works slower and uses more memory
photo_2019-07-18_15-48-43
here  benchmarks for different types of checked strings with different check ways

Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: I'd gladly accept a couple of milliseconds overhead if that means code is simpler and more readable. Your current implementation is O(N). Regex.IsMatch() is probably in the same order or at most O(N^2). But N is too small in this case to make a significant difference in any case. So I still prefer the Regex solution here.

**Comment by RannyRanny** [comment]
@hiranya911 Added requested changes

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
+using System;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs]: Add license header.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+            Assert.Throws<ArgumentException>(() => AnalyticsLabelChecker.ValidateAnalyticsLabel(string.Empty));
+        }
+
+        [Fact]
+        public void AnalyticsLabelInvalidCharacters()
+        {
+            Assert.Throws<ArgumentException>(() => AnalyticsLabelChecker.ValidateAnalyticsLabel("label(label)"));
+        }
+    }
+}
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs]: Add newline.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+            var options = new AndroidFcmOptions() { AnalyticsLabel = "label" };
+            var result = options.CopyAndValidate();
+            Assert.Equal(options.AnalyticsLabel, result.AnalyticsLabel);
+        }
+
+        [Fact]
+        public void FcmOptionsCopyAndValidateNullLabel()
+        {
+            var options = new FcmOptions() { AnalyticsLabel = null };
+            Assert.Throws<ArgumentException>(() => options.CopyAndValidate());
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs]: I don't think this should throw.

File: FirebaseAdmin/FirebaseAdmin/Messaging/ApnsFcmOptions.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
+using FirebaseAdmin.Messaging.Util;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/ApnsFcmOptions.cs]: License header

File: FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
+using FirebaseAdmin.Messaging.Util;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/FcmOptions.cs]: License header

File: FirebaseAdmin/FirebaseAdmin/Messaging/AndroidFcmOptions.cs [hunk 20]
```diff
@@ -0,0 +1,32 @@
+using FirebaseAdmin.Messaging.Util;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/AndroidFcmOptions.cs]: License header

File: FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        private static string pattern = "^[a-zA-Z0-9-_.~%]{0,50}$";
+
+        /// <summary>
+        /// Checks anytytics labels an throw if not valid.
+        /// </summary>
+        /// <exception cref="ArgumentException">If analytics label does not match pattern.</exception>
+        /// <param name="analyticsLabel">Analytics label.</param>
+        public static void ValidateAnalyticsLabel(string analyticsLabel)
+        {
+            if (string.IsNullOrWhiteSpace(analyticsLabel))
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/Util/AnalyticsLabelChecker.cs]: I think we ought to allow null values for AnalyticsLabel. In the future we will have other properties in `FcmOptions`, and it should be possible to specify them and leave AnalyticsLabel as null.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        [Fact]
+        public void AnalyticsLabelTooLong()
+        {
+            Assert.Throws<ArgumentException>(() => AnalyticsLabelChecker.ValidateAnalyticsLabel("012345678901234567890123456789012345678901234567890"));
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FcmOptionsTest.cs]: `new String('a', 51)` to easily construct a string longer than 50.

**Comment by hiranya911** [comment]
@egilmorez can you look at the documentation bits of this PR?

**Comment by RannyRanny** [comment]
@hiranya911 This is my first PR for open source. Can U please say when this will be available to use in NuGet package?

**Comment by hiranya911** [comment]
@RannyRanny most likely next week.
